File: src/steering_functions/src/hc_cc_state_space/hc_cc_state_space.cpp
/*********************************************************************
 *  Copyright (c) 2017 Robert Bosch GmbH.
 *  All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 ***********************************************************************/

#include "steering_functions/hc_cc_state_space/hc_cc_state_space.hpp"

#include <algorithm>
#include <cmath>
#include <iterator>

#include "steering_functions/utilities/utilities.hpp"

using namespace std;

namespace steering
{

    HC_CC_State_Space::HC_CC_State_Space(double kappa, double sigma, double discretization)
        : kappa_(kappa), sigma_(sigma), discretization_(discretization)
    {
        // assert positive inputs
        assert(kappa > 0.0 && sigma > 0.0 && discretization > 0.0);
        // intermediate configuration after first clothoid
        double length_min = kappa / sigma;
        double x_i, y_i, theta_i;
        if (length_min > get_epsilon())
        {
            double kappa_i;
            end_of_clothoid(0, 0, 0, 0, sigma, 1, length_min, &x_i, &y_i, &theta_i, &kappa_i);
        }
        else
        {
            x_i     = 0;
            y_i     = 0;
            theta_i = 0;
        }
        // radius
        double xc, yc;
        xc            = x_i - sin(theta_i) / kappa;
        yc            = y_i + cos(theta_i) / kappa;
        double radius = point_distance(xc, yc, 0.0, 0.0);
        // mu
        double mu     = atan(fabs(xc / yc));
        double sin_mu = sin(mu);
        double cos_mu = cos(mu);
        // delta_min
        double delta_min = 0.5 * pow(kappa, 2) / sigma;
        // assign
        hc_cc_circle_param_.set_param(kappa, sigma, radius, mu, sin_mu, cos_mu, delta_min);
    }

    void HC_CC_State_Space::set_filter_parameters(const Motion_Noise&      motion_noise,
                                                  const Measurement_Noise& measurement_noise,
                                                  const Controller&        controller)
    {
        ekf_.set_parameters(motion_noise, measurement_noise, controller);
    }

    vector<State> HC_CC_State_Space::get_path(const State&          state1,
                                              const State&          state2,
                                              std::vector<Control>& controls)
    {
        auto un_filtered_controls = get_controls(state1, state2);

        for (auto& control : un_filtered_controls)
        {
            if (abs(control.delta_s) > 1e-6)
            {
                controls.push_back(control);
            }
        }

        return integrate(state1, controls);
    }

    vector<State_With_Covariance>
    HC_CC_State_Space::get_path_with_covariance(const State_With_Covariance& state1,
                                                const State&                 state2) const
    {
        vector<Control> controls = get_controls(state1.state, state2);
        return integrate_with_covariance(state1, controls);
    }

    vector<State> HC_CC_State_Space::integrate(const State&           state,
                                               const vector<Control>& controls) const
    {
        return integrate(state, controls, discretization_);
    }

    vector<State> HC_CC_State_Space::integrate(const State&           state,
                                               const vector<Control>& controls,
                                               double                 discretization_) const
    {
        vector<State> path;
        if (controls.size() < 1)
        {
            return path;
        }

        State state_curr{}, state_next;
        // reserve capacity of path
        int n_states(0);
        for (const auto& control : controls)
        {
            double abs_delta_s(fabs(control.delta_s));
            n_states += ceil(abs_delta_s / discretization_);
        }
        path.reserve(n_states + 13);
        // push back first state
        state_curr.x     = state.x;
        state_curr.y     = state.y;
        state_curr.theta = state.theta;
        state_curr.kappa = controls.front().kappa;
        state_curr.sigma = controls.front().sigma;
        state_curr.d     = sgn(controls.front().delta_s);
        path.push_back(state_curr);

        for (const auto& control : controls)
        {
            double delta_s(control.delta_s);
            double abs_delta_s(fabs(delta_s));
            double kappa(control.kappa);
            double s_seg(0.0);
            double integration_step(0.0);

            int    n              = max(2.0, ceil(abs_delta_s / discretization_));
            double discretization = abs_delta_s / n;
            assert(discretization > 0);

            for (int i = 0; i < n; ++i)
            {
                // get integration step
                s_seg += discretization;
                if (s_seg > abs_delta_s)
                {
                    integration_step = discretization - (s_seg - abs_delta_s);
                    s_seg            = abs_delta_s;
                }
                else
                {
                    integration_step = discretization;
                }
                state_next = integrate_ODE(state_curr, control, integration_step);
                path.push_back(state_next);
                state_curr = state_next;
            }
        }
        return path;
    }

    vector<State_With_Covariance>
    HC_CC_State_Space::integrate_with_covariance(const State_With_Covariance& state,
                                                 const vector<Control>&       controls) const
    {
        vector<State_With_Covariance> path_with_covariance;
        State_With_Covariance         state_curr, state_pred, state_next;
        // reserve capacity of path
        int n_states(0);
        for (const auto& control : controls)
        {
            double abs_delta_s(fabs(control.delta_s));
            n_states += ceil(abs_delta_s / discretization_);
        }
        path_with_covariance.reserve(n_states + 3);
        // push back first state
        state_curr.state.x     = state.state.x;
        state_curr.state.y     = state.state.y;
        state_curr.state.theta = state.state.theta;
        state_curr.state.kappa = controls.front().kappa;
        state_curr.state.d     = sgn(controls.front().delta_s);
        std::copy(std::begin(state.Sigma), std::end(state.Sigma), state_curr.Sigma);
        std::copy(std::begin(state.Lambda), std::end(state.Lambda), state_curr.Lambda);
        std::copy(std::begin(state.covariance), std::end(state.covariance), state_curr.covariance);
        path_with_covariance.push_back(state_curr);

        for (const auto& control : controls)
        {
            double delta_s(control.delta_s);
            double abs_delta_s(fabs(delta_s));
            double kappa(control.kappa);
            double s_seg(0.0);
            double integration_step(0.0);
            // push_back current state if curvature discontinuity
            if (fabs(kappa - state_curr.state.kappa) > get_epsilon())
            {
                state_curr.state.kappa = kappa;
                state_curr.state.d     = sgn(delta_s);
                path_with_covariance.push_back(state_curr);
            }

            for (int i = 0, n = ceil(abs_delta_s / discretization_); i < n; ++i)
            {
                // get integration step
                s_seg += discretization_;
                if (s_seg > abs_delta_s)
                {
                    integration_step = discretization_ - (s_seg - abs_delta_s);
                    s_seg            = abs_delta_s;
                }
                else
                {
                    integration_step = discretization_;
                }
                // predict
                state_pred.state = integrate_ODE(state_curr.state, control, integration_step);
                ekf_.predict(state_curr, control, integration_step, state_pred);
                // update
                state_next.state = state_pred.state;
                ekf_.update(state_pred, state_next);

                path_with_covariance.push_back(state_next);
                state_curr.state = state_next.state;
                std::copy(std::begin(state_next.Sigma), std::end(state_next.Sigma), state_curr.Sigma);
                std::copy(
                    std::begin(state_next.Lambda), std::end(state_next.Lambda), state_curr.Lambda);
                std::copy(std::begin(state_next.covariance),
                          std::end(state_next.covariance),
                          state_curr.covariance);
            }
        }
        return path_with_covariance;
    }

    State HC_CC_State_Space::interpolate(const State&           state,
                                         const vector<Control>& controls,
                                         double                 t) const
    {
        State state_curr{state}, state_inter{state};

        // get arc length at t
        double s_path(0.0);
        double s_inter(0.0);
        for (const auto& control : controls)
        {
            s_path += fabs(control.delta_s);
        }
        t       = min(max(t, 0.0), 1.0);
        s_inter = t * s_path;

        double s(0.0);

        for (const auto& control : controls)
        {
            double delta_s(control.delta_s);
            double abs_delta_s(fabs(delta_s));

            if (s_inter - s > abs_delta_s)
            {
                state_curr = integrate_ODE(state_curr, control, abs_delta_s);
                s += abs_delta_s;
            }
            else
            {
                state_inter = integrate_ODE(state_curr, control, s_inter - s);
                break;
            }
        }
        return state_inter;
    }

    std::vector<std::vector<Control>> HC_CC_State_Space::get_all_controls(const State& state1,
                                                                          const State& state2) const
    {
        // TODO Not implemented yet;
        std::vector<std::vector<Control>> all_controls;
        auto                              control = get_controls(state1, state2);
        all_controls.push_back(control);

        return all_controls;
    }

    inline State HC_CC_State_Space::integrate_ODE(const State&   state,
                                                  const Control& control,
                                                  double         integration_step) const
    {
        State  state_next = {};
        double kappa(control.kappa);
        double sigma(control.sigma);
        double d(sgn(control.delta_s));
        if (fabs(sigma) > get_epsilon())
        {
            end_of_clothoid(state.x,
                            state.y,
                            state.theta,
                            state.kappa,
                            sigma,
                            d,
                            integration_step,
                            &state_next.x,
                            &state_next.y,
                            &state_next.theta,
                            &state_next.kappa);
            state_next.sigma = sigma;
        }
        else
        {
            if (fabs(kappa) > get_epsilon())
            {
                end_of_circular_arc(state.x,
                                    state.y,
                                    state.theta,
                                    kappa,
                                    d,
                                    integration_step,
                                    &state_next.x,
                                    &state_next.y,
                                    &state_next.theta);
                state_next.kappa = kappa;
                // state_next.sigma = 0.0;
            }
            else
            {
                end_of_straight_line(state.x,
                                     state.y,
                                     state.theta,
                                     d,
                                     integration_step,
                                     &state_next.x,
                                     &state_next.y);
                state_next.theta = state.theta;
                // state_next.kappa = state.kappa;
            }
        }
        state_next.d = d;
        state_next.s = state.s + integration_step;
        return state_next;
    }

} // namespace steering

File: rust/src/hc_cc_state_space/hc_cc_state_space.rs
// Copyright (c) 2017 Robert Bosch GmbH.
// All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::hc_cc_circle::HcCcCircleParam;
use crate::base_state_space::StateSpace;
use crate::state::{Control, State};
use crate::utilities::{end_of_clothoid, get_epsilon, point_distance};

/// Base data for HC/CC state spaces.
///
/// This struct holds the common parameters (curvature, sharpness,
/// discretisation step, and the derived circle parameters) shared by all
/// HC/CC steering variants.  Concrete state spaces (CC-Dubins, HC-RS,
/// CC-RS, …) embed an `HcCcStateSpace` and implement the [`StateSpace`]
/// trait by delegating to its helper methods.
#[derive(Debug, Clone)]
pub struct HcCcStateSpace {
    /// Maximum curvature.
    pub kappa: f64,
    /// Maximum sharpness.
    pub sigma: f64,
    /// Discretisation step for path integration.
    pub discretization: f64,
    /// Derived HC/CC circle parameters.
    pub hc_cc_circle_param: HcCcCircleParam,
}

impl HcCcStateSpace {
    /// Create a new `HcCcStateSpace`.
    ///
    /// # Panics
    ///
    /// Panics if `kappa`, `sigma`, or `discretization` is not positive.
    pub fn new(kappa: f64, sigma: f64, discretization: f64) -> Self {
        assert!(
            kappa > 0.0 && sigma > 0.0 && discretization > 0.0,
            "kappa, sigma and discretization must be positive"
        );

        // Intermediate configuration after the first clothoid.
        let length_min = kappa / sigma;
        let (x_i, y_i, theta_i) = if length_min > get_epsilon() {
            let (x, y, th, _kappa) = end_of_clothoid(0.0, 0.0, 0.0, 0.0, sigma, 1.0, length_min);
            (x, y, th)
        } else {
            (0.0, 0.0, 0.0)
        };

        // Centre of the circle tangent to the clothoid end-point.
        let xc = x_i - theta_i.sin() / kappa;
        let yc = y_i + theta_i.cos() / kappa;
        let radius = point_distance(xc, yc, 0.0, 0.0);

        let mu = (xc / yc).abs().atan();
        let sin_mu = mu.sin();
        let cos_mu = mu.cos();
        let delta_min = 0.5 * kappa * kappa / sigma;

        let mut hc_cc_circle_param = HcCcCircleParam::default();
        hc_cc_circle_param.set_param(kappa, sigma, radius, mu, sin_mu, cos_mu, delta_min);

        Self {
            kappa,
            sigma,
            discretization,
            hc_cc_circle_param,
        }
    }

    /// Filter negligible control segments and integrate.
    ///
    /// This is the HC/CC override of `get_path`: it obtains unfiltered
    /// controls via `raw_controls`, removes segments with
    /// `|delta_s| ≤ 1e-6`, and integrates the remainder.
    pub fn get_path_filtered<F>(
        &self,
        state1: &State,
        state2: &State,
        raw_controls_fn: F,
    ) -> (Vec<State>, Vec<Control>)
    where
        F: FnOnce(&State, &State) -> Vec<Control>,
    {
        let unfiltered = raw_controls_fn(state1, state2);
        let controls: Vec<Control> = unfiltered
            .into_iter()
            .filter(|c| c.delta_s.abs() > 1e-6)
            .collect();
        let path = self.integrate_controls(state1, &controls);
        (path, controls)
    }

    /// Integrate a control sequence using the stored discretisation.
    ///
    /// This is a convenience wrapper so that embedders don't have to
    /// construct a full [`StateSpace`] implementor just to call `integrate`.
    fn integrate_controls(&self, state: &State, controls: &[Control]) -> Vec<State> {
        /// A thin adapter that exposes `self.discretization` via the trait.
        struct Adapter(f64);
        impl StateSpace for Adapter {
            fn discretization(&self) -> f64 {
                self.0
            }
            fn get_controls(&self, _: &State, _: &State) -> Vec<Control> {
                Vec::new()
            }
            fn get_all_controls(&self, _: &State, _: &State) -> Vec<Vec<Control>> {
                Vec::new()
            }
        }
        let adapter = Adapter(self.discretization);
        adapter.integrate(state, controls, self.discretization)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_basic() {
        let ss = HcCcStateSpace::new(1.0, 1.0, 0.1);
        assert!((ss.kappa - 1.0).abs() < 1e-12);
        assert!((ss.sigma - 1.0).abs() < 1e-12);
        assert!((ss.discretization - 0.1).abs() < 1e-12);
        assert!(ss.hc_cc_circle_param.radius > 0.0);
        assert!(ss.hc_cc_circle_param.delta_min > 0.0);
    }

    #[test]
    #[should_panic]
    fn test_new_zero_kappa() {
        HcCcStateSpace::new(0.0, 1.0, 0.1);
    }

    #[test]
    #[should_panic]
    fn test_new_negative_sigma() {
        HcCcStateSpace::new(1.0, -1.0, 0.1);
    }

    #[test]
    fn test_param_consistency() {
        let ss = HcCcStateSpace::new(2.0, 1.0, 0.1);
        let p = &ss.hc_cc_circle_param;
        assert!((p.kappa - 2.0).abs() < 1e-12);
        assert!((p.sigma - 1.0).abs() < 1e-12);
        assert!((p.kappa_inv - 0.5).abs() < 1e-12);
        assert!((p.delta_min - 2.0).abs() < 1e-12); // 0.5 * 4.0 / 1.0
    }

    #[test]
    fn test_get_path_filtered_removes_tiny() {
        let ss = HcCcStateSpace::new(1.0, 1.0, 0.1);
        let s1 = State::new(0.0, 0.0, 0.0, 0.0);
        let s2 = State::new(1.0, 0.0, 0.0, 0.0);
        let (path, controls) = ss.get_path_filtered(&s1, &s2, |_, _| {
            vec![
                Control {
                    delta_s: 1e-8,
                    kappa: 0.0,
                    sigma: 0.0,
                },
                Control {
                    delta_s: 1.0,
                    kappa: 0.0,
                    sigma: 0.0,
                },
            ]
        });
        // The tiny segment should be filtered out
        assert_eq!(controls.len(), 1);
        assert!(!path.is_empty());
    }
}
